src/moonlink_metadata_store/src/error.rs

```diff
@@ -8,65 +9,77 @@ use tokio_postgres::Error as TokioPostgresError;
 #[derive(Clone, Debug, Error)]
 pub enum Error {
     #[cfg(feature = "metadata-postgres")]
-    #[error("tokio postgres error: {source}")]
-    TokioPostgres { source: Arc<TokioPostgresError> },
+    #[error("{0}")]
+    TokioPostgres(ErrorStruct),
 
-    #[error("metadata operation row number doesn't match {0} vs {1}")]
-    PostgresRowCountError(u32, u32),
+    #[error("{0}")]
+    PostgresRowCountError(ErrorStruct),
 
-    #[error("sqlx error: {source}")]
-    Sqlx { source: Arc<sqlx::Error> },
+    #[error("{0}")]
+    Sqlx(ErrorStruct),
 
-    #[error("metadata operation row number doesn't match {0} vs {1}")]
-    SqliteRowCountError(u32, u32),
+    #[error("{0}")]
+    SqliteRowCountError(ErrorStruct),
 
-    #[error("metadata access failed precondition: {0}")]
-    MetadataStoreFailedPrecondition(String),
+    #[error("{0}")]
+    MetadataStoreFailedPrecondition(ErrorStruct),
 
-    #[error("serde json error: {source}")]
-    SerdeJson { source: Arc<SerdeJsonError> },
+    #[error("{0}")]
+    SerdeJson(ErrorStruct),
 
-    #[error("table id with id {0} not found")]
-    TableIdNotFound(u32),
+    #[error("{0}")]
+    TableIdNotFound(ErrorStruct),
 
-    #[error("required field {0} not exist in serialized config json")]
-    ConfigFieldNotExist(String),
+    #[error("{0}")]
+    ConfigFieldNotExist(ErrorStruct),
 
-    #[error("IO error: {source}")]
-    Io { source: Arc<std::io::Error> },
+    #[error("{0}")]
+    Io(ErrorStruct),
 }
 
 pub type Result<T> = std::result::Result<T, Error>;
 
 #[cfg(feature = "metadata-postgres")]
 impl From<TokioPostgresError> for Error {
+    #[track_caller]
     fn from(source: TokioPostgresError) -> Self {
-        Error::TokioPostgres {
-            source: Arc::new(source),
-        }
+        Error::TokioPostgres(
+            ErrorStruct::new(
+                format!("tokio postgres error: {source}"),
+                ErrorStatus::Permanent,
+            )
+            .with_source(source),
+        )
     }
 }
 
 impl From<sqlx::Error> for Error {
+    #[track_caller]
     fn from(source: sqlx::Error) -> Self {
-        Error::Sqlx {
-            source: Arc::new(source),
-        }
+        Error::Sqlx(
+            ErrorStruct::new(format!("sqlx error: {source}"), ErrorStatus::Permanent)
+                .with_source(source),
+        )
     }
 }
 
 impl From<SerdeJsonError> for Error {
+    #[track_caller]
     fn from(source: SerdeJsonError) -> Self {
-        Error::SerdeJson {
-            source: Arc::new(source),
-        }
+        Error::SerdeJson(
+            ErrorStruct::new(
+                format!("serde json error: {source}"),
```

Comment:
I think we're duplicating `source` information both at string format message and in `source` field, do you mind keeping it only at `source` field? Same for other errors